### PR TITLE
vpmu:cpu cycles checking between host and guest

### PR DESCRIPTION
**Pull request by maxujun**
ID:1937222
Signed-off-by: Xujun Ma <xuma@redhat.com>

### CODE REVIEW

File: qemu/tests/vpmu_check_instructions.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from avocado.utils import process
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    vpum cpu cycles checking between host and guest:
+    1) boot guest
+    2) check cpu cycles for host
+    3) check cpu cycles for guest and compare wirh host
```
Comment (MiriamDeng) [inline qemu/tests/vpmu_check_instructions.py:18]: it's typo mistake "wirh"

File: qemu/tests/vpmu_check_instructions.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    if process.system("test -x perf", ignore_status=True, shell=True) != 0:
+        process.getstatusoutput("yum install -y perf")
+
+    error_context.context("build binary file 'million' in host", logging.info)
+    os.chdir(src_dir)
+    process.getstatusoutput(build_cmd % host_arch)
+
+    error_context.context("running test command in host", logging.info)
+    s, o = process.getstatusoutput(test_cmd)
+
```
Comment (MiriamDeng) [inline qemu/tests/vpmu_check_instructions.py:47]: it seems you need to check the status of the command also

File: qemu/tests/vpmu_check_instructions.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    dst_file = os.path.join(tmp_dir, "million-%s.s" % host_arch)
+    if process.system("test -x perf", ignore_status=True, shell=True) != 0:
+        process.getstatusoutput("yum install -y perf")
+
+    error_context.context("build binary file 'million' in host", logging.info)
+    os.chdir(src_dir)
+    process.getstatusoutput(build_cmd % host_arch)
+
+    error_context.context("running test command in host", logging.info)
+    s, o = process.getstatusoutput(test_cmd)
```
Comment (MiriamDeng) [inline qemu/tests/vpmu_check_instructions.py:46]: it seems, it returned "status" but you didn't check it

Comment (MiriamDeng) [reply qemu/tests/vpmu_check_instructions.py:45]: A reference for L40 to L45
cmds.append()
for cmd in cmds:
s, o = process.getstatusoutput(cmd)
if s:
test.error/fail("")

File: qemu/tests/vpmu_check_instructions.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import os
+import logging
+import re
+import platform
+
+from virttest import data_dir
+from virttest import error_context
+from avocado.utils import process
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:8]: Please import `avocado` module before `virttest`

File: qemu/tests/vpmu_check_instructions.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    host_arch = platform.machine()
+    src_dir = os.path.join(data_dir.get_deps_dir(), 'million')
+    src_file = os.path.join(src_dir, "million-%s.s" % host_arch)
+    dst_file = os.path.join(tmp_dir, "million-%s.s" % host_arch)
+    if process.system("test -x perf", ignore_status=True, shell=True) != 0:
+        process.getstatusoutput("yum install -y perf")
+
+    error_context.context("build binary file 'million' in host", logging.info)
+    os.chdir(src_dir)
+    process.getstatusoutput(build_cmd % host_arch)
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:42]: Suggest remove this line and build the test suite with an absolute path. Copy the `million` file into tmp dir and build it.

File: qemu/tests/vpmu_check_instructions.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    error_context.context("running test command in host", logging.info)
+    s, o = process.getstatusoutput(test_cmd)
+
+    host_cpu_cycles = re.findall(r"(\d+) *instructions:u", o, re.M)
+
+    if session.cmd_status("test -x perf"):
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:49]: Also, can use 
```
utils_package.package_install("perf", session)
```

File: qemu/tests/cfg/vpmu_check.cfg [hunk 7]
```diff
@@ -0,0 +1,8 @@
+- vpmu_check:
+    type = vpmu_check_instructions
+    start_vm = yes
+    only Linux
```
Comment (PaulYuuu) [inline qemu/tests/cfg/vpmu_check.cfg:4]:
```suggestion
    only Linux
    only ppc64, ppc64le
```

**Comment by maxujun** [comment]
Updated

File: qemu/tests/vpmu_check_instructions.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import os
+import logging
+import re
+
+from avocado.utils import process
+from virttest import data_dir
+from virttest import error_context
+from virttest import arch
+from virttest import utils_package
+from shutil import copyfile
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:10]: https://github.com/autotest/tp-qemu/blob/46767e4ef7321939decd81d64dc754adf20391dc/CODING_STYLE#L44-L50

File: qemu/tests/vpmu_check_instructions.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    :param test: QEMU test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    check_timeout = int(params.get("check_timeout", "600"))
+    tmp_dir = params.get("tmp_dir")
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    timeout = int(params.get("login_timeout", 360))
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:30]: `params.get_numeric`

File: qemu/tests/vpmu_check_instructions.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    vm_arch = params["vm_arch_name"]
+    host_arch = arch.ARCH
+    src_dir = os.path.join(data_dir.get_deps_dir(), 'million')
+    src_file = os.path.join(src_dir, "million-%s.s" % host_arch)
+    dst_file = os.path.join(tmp_dir, "million-%s.s" % host_arch)
+    utils_package.package_install("perf")
+
+    error_context.context("build binary file 'million' in host", logging.info)
+    copyfile(src_file, dst_file)
+    process.getstatusoutput(build_cmd % host_arch)
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:43]: `process.run` if you don't need status and output.

File: qemu/tests/vpmu_check_instructions.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    s, o = process.getstatusoutput(test_cmd)
+    if s:
+        test.fail("Failed to run test command")
+
+    host_cpu_cycles = re.findall(r"(\d+) *instructions:u", o, re.M)
+
+    utils_package.package_install("perf", session)
+    error_context.context("transfer '%s' to guest('%s')" %
+                          (src_file, dst_file), logging.info)
+    vm.copy_files_to(src_file, tmp_dir, timeout=120)
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:55]: Should we copy the file with guest arch(vm_arch)? The `src_file` is more like for the host only.

Comment (maxujun) [reply qemu/tests/vpmu_check_instructions.py:55]: What do you mean? I copy host file to guest  with vm.copy_files_to.

Comment (PaulYuuu) [reply qemu/tests/vpmu_check_instructions.py:55]: `src_file = os.path.join(src_dir, "million-%s.s" % host_arch)`
If the host is ppc64le and the guest is ppc64.

**Comment by maxujun** [comment]
Updated

File: qemu/tests/vpmu_check_instructions.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    s, o = process.getstatusoutput(test_cmd)
+    if s:
+        test.fail("Failed to run test command")
+
+    host_cpu_cycles = re.findall(r"(\d+) *instructions:u", o, re.M)
+
+    utils_package.package_install("perf", session)
+    error_context.context("transfer '%s' to guest('%s')" %
+                          (src_file, dst_file), logging.info)
+    src_file = os.path.join(src_dir, "million-%s.s" % vm_arch)
```
Comment (PaulYuuu) [inline qemu/tests/vpmu_check_instructions.py:57]: This line should be moved above L55, to let `error_context.context` display a correct `src_file`.
Other LGTM.

**Comment by maxujun** [comment]
Updated

**Review by PaulYuuu — APPROVED** [review]

**Comment by PaulYuuu** [comment]
@MiriamDeng Any new comments?

File: qemu/tests/vpmu_check_instructions.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    vpum cpu cycles checking between host and guest:
+    1) boot guest
+    2) check cpu cycles for host
+    3) check cpu cycles for guest and compare with host
+
+    :param test: QEMU test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    check_timeout = int(params.get("check_timeout", "600"))
```
Comment (MiriamDeng) [inline qemu/tests/vpmu_check_instructions.py:27]: Check_timeout = 300 in cfg, they should be the same, right ?

File: qemu/tests/vpmu_check_instructions.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    timeout = params.get_numeric("login_timeout", 360)
+    session = vm.wait_for_login(timeout=timeout)
+    test_cmd = params.get("test_cmd")
+    build_cmd = params.get("build_cmd")
+    vm_arch = params["vm_arch_name"]
+    host_arch = arch.ARCH
+    src_dir = os.path.join(data_dir.get_deps_dir(), 'million')
+    src_file = os.path.join(src_dir, "million-%s.s" % host_arch)
+    dst_file = os.path.join(tmp_dir, "million-%s.s" % host_arch)
+    utils_package.package_install("perf")
```
Comment (MiriamDeng) [inline qemu/tests/vpmu_check_instructions.py:41]: In my opinions, we need to check the installation status
if not utils_package.package_install(pkgs, session):
            test.error("Install dependency packages failed")

File: qemu/tests/vpmu_check_instructions.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    process.run(build_cmd % host_arch)
+
+    error_context.context("running test command in host", logging.info)
+    s, o = process.getstatusoutput(test_cmd)
+    if s:
+        test.fail("Failed to run test command")
+
+    host_cpu_cycles = re.findall(r"(\d+) *instructions:u", o, re.M)
+
+    utils_package.package_install("perf", session)
```
Comment (MiriamDeng) [inline qemu/tests/vpmu_check_instructions.py:54]: In my opinions, we need to check the installation status
if not utils_package.package_install(pkgs, session):
test.error("Install dependency packages failed")

Comment (MiriamDeng) [reply qemu/tests/vpmu_check_instructions.py:54]: Also has the following problem while applying it 
[root@ibm-p9b-12 io-github-autotest-qemu]# git am 2678.patch
Applying: vpmu:cpu cycles checking between host and guest
.git/rebase-apply/patch:21: space before tab in indent.
     	     # count for 1 million instructions
.git/rebase-apply/patch:51: space before tab in indent.
     	     # count for 1 million instructions
warning: 2 lines add whitespace errors.

**Comment by maxujun** [comment]
updated

**Comment by MiriamDeng** [comment]
@maxujun ,after applying it, still can't filter the test case, could you double check it ?

**Pull request merged**